#include <sys/time.h>
#include <sys/resource.h>
#include <unistd.h>
#include <sys/wait.h>
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <math.h>
#include <limits.h>
#include <signal.h>
#include <pthread.h>
#include <stdatomic.h>
#include <stdbool.h>
#include "misc.h"
#include "tinygraph.h"
#include "graph.h"
#include "heap.h"
#include "blant.h"
#include "queue.h"
#include "multisets.h"
#include "sorts.h"
#include "combin.h"
#include "blant-window.h"
#include "blant-output.h"
#include "blant-utils.h"
#include "blant-sampling.h"
#include "blant-synth-graph.h"
#include "rand48.h"
Boolean _earlyAbort; // Can be set true by anybody anywhere, and they're responsible for producing a warning as to why
#include "blant-predict.h"
#include "importance.h"
#include "odv.h"
#include "stats.h"

// define random variable algorithm
#define USE_MarsenneTwister 0
#if USE_MarsenneTwister
#include "../C++/mt19937.h" // not yet implemented correctly
_Thread_local MT19937 *mt19937Seed;
void RandomSeed(long seed) {
    if(mt19937Seed) Mt19937Free(mt19937Seed);
    mt19937Seed = Mt19937Alloc(seed);
}
double RandomUniform(void) {
    return Mt19937NextDouble(mt19937Seed);
}
#else
#include "rand48.h"
_Thread_local unsigned short erand48Seed[3];
void RandomSeed(long seed) {
    // split the number seed into 3 portions of 16 bits
    erand48Seed[0] = (unsigned short)(seed & 0xFFFF);         // Lower 16 bits
    erand48Seed[1] = (unsigned short)((seed >> 16) & 0xFFFF); // Middle 16 bits
    erand48Seed[2] = 0x330E; // Upper 16 bits, set to this to mimic srand48
}
double RandomUniform(void) {
    return erand48(erand48Seed);
}
#endif


static int _numNodes, _numEdges, _maxEdges=1024, _seed = -1; // -1 means "not initialized"
static unsigned *_pairs;
static float *_weights;
char **_nodeNames, _supportNodeNames = true;
static FILE *interestFile;
Boolean _child; // are we a child process?
int _quiet; // suppress notes and/or warnings, higher value = more quiet

char * _sampleFileName;

// _k is the global variable storing k; _Bk=actual number of entries in the canon_map for given k.
unsigned int _k, _min_edge_count;
unsigned int _Bk, _k_small;

unsigned long _known_canonical_count[] =
	{0, 1, 2, 4, 11, 34, 156, 1044, 12346, 274668, 12005168, 1018997864, 165091172592}; // note (k=12) too big for 32 bits
	//k=1  2  3   4   5   6     7     8      9        10         11          12

Gint_type _alphaList[MAX_CANONICALS];
Gordinal_type _numCanon;
char _canonNumEdges[MAX_CANONICALS];
double _totalStarMotifs; // This is a double because the value can *way* overflow even a 128-bit integer on large graphs.
int _canonNumStarMotifs[MAX_CANONICALS]; // However, the per-canonical values are integers
Gint_type _canonList[MAX_CANONICALS]; // map ordinals to integer representation of the canonical
SET *_connectedCanonicals; // the SET of canonicals that are connected.
SET ***_communityNeighbors;
char _communityMode; // 'g' for graphlet or 'o' for orbit
Boolean _useComplement; // to use the complement graph (DEPRECATED, FIND ANOTHER LETTER)
Boolean _weighted; // input network is weighted
Boolean _rawCounts;
Gordinal_type _numConnectedCanon;
int _numConnectedComponents;
int *_componentSize;
int *_startNodes, _numStartNodes;
SET *_startNodeSet;

Gint_type _numOrbits, _orbitList[MAX_CANONICALS][MAX_K]; // map from [ordinal][canonicalNode] to orbit ID.
Gordinal_type _orbitCanonMapping[MAX_ORBITS]; // Maps orbits to canonical (including disconnected)
char _orbitCanonNodeMapping[MAX_ORBITS]; // Maps orbits to canonical (including disconnected)
int *_whichComponent;

enum OutputMode _outputMode = undef;
unsigned long _numSamples;
int **_graphletDistributionTable;
double _graphletCount[MAX_CANONICALS], _graphletConcentration[MAX_CANONICALS], _absoluteCountMultiplier=1;
unsigned long _batchRawCount[MAX_CANONICALS], _batchRawTotalSamples; // batches for confidence intervals

enum CanonicalDisplayMode _displayMode = undefined;
enum FrequencyDisplayMode _freqDisplayMode = freq_display_mode_undef;

int _outputMapping[MAX_CANONICALS];

int _orca_orbit_mapping[58];
int _connectedOrbits[MAX_ORBITS];
int _numConnectedOrbits;

// A bit counter-intuitive: we need to allocate this many vectors each of length [_numNodes],
// and then the degree for node v, graphlet/orbit g is _degreeVector[g][v], NOT [v][g].
// We do this simply because we know the length of MAX_CANONICALS so we pre-know the length of
// the first dimension, otherwise we'd need to get more funky with the pointer allocation.
// Only one of these actually get allocated, depending upon outputMode.
double *_graphletDegreeVector[MAX_CANONICALS];
double *_orbitDegreeVector[MAX_ORBITS];

double *_cumulativeProb, _worstPrecision, _meanPrec;
enum PrecisionMode _precisionMode = mean;
enum PrecisionWeights _precisionWt = PrecWtNone;
int _worstCanon=-1;

double _desiredDigits, _desiredPrec, _confidence;


// Here's the actual mapping from non-canonical to canonical, same argument as above wasting memory, and also mmap'd.
// So here we are allocating 256MB x sizeof(short int) = 512MB.
// Grand total statically allocated memory is exactly 1.25GB.
//static short int _K[maxBk] __attribute__ ((aligned (8192)));
Gordinal_type *_K = NULL; // Allocating memory dynamically

/* AND NOW THE CODE */

static unsigned int **_componentList; // list of lists of components, largest to smallest.
static double _totalCombinations, *_combinations, *_probOfComponent;
SET **_componentSet;

// number of parallel threads required, and the maximum allowed at one time (set to sysconf(_SC_NPROCESSORS_ONLN) later in main).
int _numThreads, _maxThreads;
_Atomic bool _doneSampling = false;
Accumulators _trashAccumulator = {};
Accumulators *_threadAccumulators;
enum StopMode _stopMode;

double GetCPUseconds(void) {
    struct rusage usg;
    double res = 0;
    if (getrusage(RUSAGE_SELF, &usg) == 0) {
        res += (usg.ru_utime.tv_sec + usg.ru_stime.tv_sec) +
               (usg.ru_utime.tv_usec + usg.ru_stime.tv_usec)/1e6;
    } else {
        return -1;
    }
    if (getrusage(RUSAGE_CHILDREN, &usg) == 0) {
        res += (usg.ru_utime.tv_sec + usg.ru_stime.tv_sec) +
               (usg.ru_utime.tv_usec + usg.ru_stime.tv_usec)/1e6;
    } else {
        return -1;
    }
    return res;
}

#define O_ALLOC 1
static int InitializeConnectedComponents(GRAPH *G)
{
    static unsigned v, *Varray, j, i;
    assert(!Varray); // we only can be called once.
    Varray = Calloc(G->n, sizeof(int));
    assert(_numConnectedComponents == 0);
    SET *visited = SetAlloc(G->n);

    // Allocate these all with Ocalloc since they never get freed
    _whichComponent = Ocalloc(G->n, sizeof(int));
    _componentSize = Ocalloc(G->n, sizeof(int)); // probably bigger than it needs to be but...
    _componentList = Ocalloc(G->n, sizeof(int*)); // probably bigger...
    _combinations = Ocalloc(G->n, sizeof(double*)); // probably bigger...
    _probOfComponent = Ocalloc(G->n, sizeof(double*)); // probably bigger...
    _cumulativeProb = Ocalloc(G->n, sizeof(double*)); // probably bigger...
    _componentSet = Ocalloc(G->n, sizeof(SET*));

    int nextStart = 0;
    _componentList[0] = Varray;
    for(v=0; v < G->n; v++) if(!SetIn(visited, v))
    {
	_componentSet[_numConnectedComponents] = SetAlloc(G->n);
	_componentList[_numConnectedComponents] = Varray + nextStart;
	GraphVisitCC(G, v, _componentSet[_numConnectedComponents], Varray + nextStart, _componentSize + _numConnectedComponents);
	SetUnion(visited, visited, _componentSet[_numConnectedComponents]);
	for(j=0; j < _componentSize[_numConnectedComponents]; j++)
	{
	    assert(_whichComponent[Varray[nextStart + j]] == 0);
	    _whichComponent[Varray[nextStart + j]] = _numConnectedComponents;
	}
	nextStart += _componentSize[_numConnectedComponents];
	++_numConnectedComponents;
    }
    assert(nextStart == G->n);

    _totalCombinations = 0.0;
    for(i=0; i< _numConnectedComponents; i++)
    {
	//find the biggest one
	int biggest = i;
	for(j=i+1; j<_numConnectedComponents;j++)
	    if(_componentSize[j] > _componentSize[biggest])
		biggest = j;
	// Now swap the biggest one into position i;
	for(j=0; j < _componentSize[biggest]; j++)
	    _whichComponent[_componentList[biggest][j]] = i;
	unsigned itmp, *pitmp;
	SET * stmp;
	itmp = _componentSize[i];
	_componentSize[i] = _componentSize[biggest];
	_componentSize[biggest] = itmp;
	pitmp = _componentList[i];
	_componentList[i] = _componentList[biggest];
	_componentList[biggest] = pitmp;
	stmp = _componentSet[i];
	_componentSet[i] = _componentSet[biggest];
	_componentSet[biggest] = stmp;
	_combinations[i] = CombinChooseDouble(_componentSize[i], _k);
	_totalCombinations += _combinations[i];
    }

    double cumulativeProb = 0.0;
    for(i=0; i< _numConnectedComponents; i++)
    {
	_probOfComponent[i] =  _combinations[i] / _totalCombinations;
	_cumulativeProb[i] = cumulativeProb + _probOfComponent[i];
	cumulativeProb = _cumulativeProb[i];
	if(cumulativeProb > 1)
	{
	    assert(cumulativeProb - 1.0 < _numConnectedComponents * 1e-15);  // allow some roundoff error
	    cumulativeProb = _cumulativeProb[i] = 1.0;
	}
	//printf("Component %d has %d nodes and probability %lf, cumulative prob %lf\n", i, _componentSize[i], _probOfComponent[i], _cumulativeProb[i]);
    }
    SetFree(visited);
    Free(Varray); // but do NOT set it to NULL, as a flag not to run this again
    return _numConnectedComponents;
}

static void InitializeStarMotifs(GRAPH *G) {
    int i;
    _totalStarMotifs = 0.0;
    for(i=0; i< G->n; i++) _totalStarMotifs += CombinChooseDouble(GraphDegree(G,i),_k-1);
    if(_totalStarMotifs==0) Warning("cannot estimate absolute graphlet count because this network has no star motifs");
    //Note("totStarMotifs is %g", _totalStarMotifs);
    for(i=0; i<_numCanon; i++) _canonNumStarMotifs[i] = -1; // 0 is a valid value so use -1 to mean "not yet initialized"
}

const char *SampleMethodStr(void) {
    switch(_sampleMethod) {
    case SAMPLE_NODE_EXPANSION: return "NBE"; break;
    case SAMPLE_SEQUENTIAL_CHAINING: return "SEC"; break;
    case SAMPLE_FAYE: return "FAYE"; break;
    case SAMPLE_EDGE_EXPANSION: return "EBE"; break;
    case SAMPLE_MCMC:
	if(_sampleSubmethod == SAMPLE_MCMC_EC) return "EDGE_COVER";
        else return (_MCMC_EVERY_EDGE ? "MCMC:E" : "MCMC");
	break;
    case SAMPLE_RESERVOIR: return "RES"; break;
    case SAMPLE_ACCEPT_REJECT: return "AR"; break;
    case SAMPLE_INDEX: return "INDEX"; break;
    case SAMPLE_FROM_FILE:
	if(_sampleFile == stdin) return "STDIN";
	else return _sampleFileName;
    }
    Fatal("SampleMethodStr: unknown _sampleMethod %d", _sampleMethod);
    return NULL;
}

Gint_type alphaListPopulate(char *BUF, Gint_type *alpha_list, int k) {
    int i;
    if(_rawCounts) { // turn off alphas by making them all 1.
	for(i=0; i<_numCanon; i++) alpha_list[i] = 1;
	return _numCanon;
    }

    if(_sampleMethod == SAMPLE_NODE_EXPANSION) {
	    sprintf(BUF, "%s/%s/alpha_list_NBE%d.txt", _BLANT_DIR, _CANON_DIR, k);
    } else if(_sampleMethod == SAMPLE_EDGE_EXPANSION) {
	    sprintf(BUF, "%s/%s/alpha_list_EBE%d.txt", _BLANT_DIR, _CANON_DIR, k);
    } else {
	    sprintf(BUF, "%s/%s/alpha_list_MCMC%d.txt", _BLANT_DIR, _CANON_DIR, k);
    }
    FILE *fp_ord=fopen(BUF, "r");
    if(!fp_ord) Fatal("cannot find %s\n", BUF);
    Gint_type numAlphas=0;
    if(1!=fscanf(fp_ord, GINT_FMT, &numAlphas) || numAlphas<=0) Fatal("alphaListPopulate: fscanf failed to read numAlphas");
#if SELF_LOOPS || !SELF_LOOPS // this should be true regardless
    assert(numAlphas == _numCanon);
#endif
    for(i=0; i<numAlphas; i++)
	if(1!=fscanf(fp_ord, GINT_FMT, &alpha_list[i]))
	    Fatal("alphaListPopulate: fscanf failed to read alpha[%d]", i);
    fclose(fp_ord);
    return numAlphas;
}

// Loads alpha values(overcounting ratios) for NBE/SEC/EBE/MCMC sampling from files
// The alpha value represents the number of ways to get that graphlet
// Concentrations are initialized to 0
void initialize(GRAPH* G, int k, unsigned long numSamples) {
    int i;

    char BUF[BUFSIZ];
    _numSamples = numSamples;
    alphaListPopulate(BUF, _alphaList, k);
    for(i = 0; i < _numCanon; i++) _graphletConcentration[i] = 0.0;
}


// Loads alpha values(overcounting ratios) for MCMC sampling from files
// The alpha value represents the number of ways to walk over that graphlet
// Global variable _MCMC_L is needed by many functions
// _MCMC_L represents the length of the sliding window in d graphlets for sampling
// Global variable _numSamples needed for the algorithm to reseed halfway through
// Concentrations are initialized to 0
void initializeMCMC(GRAPH* G, int k, unsigned long numSamples) {
	_MCMC_L = k - mcmc_d  + 1;
	// Count the number of valid edges to start from
	int i, validEdgeCount = 0;
	for (i = 0; i < G->numEdges; i++)
	    if (_componentSize[_whichComponent[G->edgeList[2*i]]] >= k)
		validEdgeCount++;
	_samplesPerEdge = (numSamples + (validEdgeCount / 2)) / validEdgeCount; // Division rounding up for samples per edge
	if(_sampleSubmethod == SAMPLE_MCMC_EC) {
	    //_samplesPerEdge = numSamples;
	    _EDGE_COVER_G = GraphCopy(G);
	}

	_numSamples = numSamples;
	if(!_window) {
		initialize(G, k, numSamples);
	}
}

// Convert the graphlet frequencies to concentrations
void finalize(GRAPH *G, unsigned long numSamples) {
    double totalConcentration = 0;
    int i, j;
    for (i = 0; i < _numCanon; i++) {
	if(_graphletConcentration[i] < 0.0 || isinf(_graphletConcentration[i]))
	    Fatal("_graphletConcentration[%d] %.15g should be neither negative nor infinite\n", i, _graphletConcentration[i]);
	totalConcentration += _graphletConcentration[i];
    }
    if(totalConcentration) for (i = 0; i < _numCanon; i++) _graphletConcentration[i] /= totalConcentration;

    if(_outputMode & outputODV)
	for(i=0;i<_numOrbits;i++)
	    for(j=0;j<G->n;j++)
		_orbitDegreeVector[i][j] *= numSamples/totalConcentration;
    if(_outputMode & outputGDV)
	for(i=0;i<_numCanon;i++)
	    for(j=0;j<G->n;j++)
		_graphletDegreeVector[i][j] *= numSamples/totalConcentration;
}

#if 0 // unused code, commented out to shut up the compiler
// return how many nodes found. If you call it with startingNode == 0 then we automatically clear the visited array
static TSET _visited;
int NumReachableNodes(TINY_GRAPH *g, int startingNode)
{
    if(startingNode == 0) TSetEmpty(_visited);
    TSetAdd(_visited,startingNode);
    unsigned j, Varray[MAX_K], numVisited = 0;
    int numNeighbors = TSetToArray(Varray, g->A[startingNode]);
    assert(numNeighbors == g->degree[startingNode]);
    for(j=0; j<numNeighbors; j++)if(!TSetIn(_visited,Varray[j])) numVisited += NumReachableNodes(g,Varray[j]);
    return 1+numVisited;
}

// Compute the degree of the state in the state graph (see Lu&Bressen)
// Given the big graph G, and a set of nodes S (|S|==k), compute the
// degree of the *state* represented by these nodes, which is:
//     Degree(S) = k*|NN2| + (k-1)*|NN1|
// where NN1 is the set of nodes one step outside S that have only 1 connection back into S,
// and   NN2 is the set of nodes one step outside S that have more than 1 connection back into S.
static int StateDegree(GRAPH *G, SET *S)
{
#if PARANOID_ASSERTS
    assert(SetCardinality(S) == _k);
#endif
    unsigned Varray[_k]; // array of elements in S
    SetToArray(Varray, S);

    SET *outSet=SetAlloc(G->n); // the set of nodes in G that are one step outside S, from *anybody* in S
    int connectCount[G->n]; // for each node in the outset, count the number of times it's connected into S
    memset(connectCount, 0, G->n * sizeof(int)); // set them all to zero

    int i, j, numOut = 0;
    for(i=0;i<_k;i++) for(j=0; j<G->degree[Varray[i]]; j++)
    {
	int neighbor = G->neighbor[Varray[i]][j];
	if(!SetIn(S, neighbor))
	{
	    SetAdd(outSet, neighbor);
	    if(connectCount[neighbor] == 0) ++numOut;
	    ++connectCount[neighbor];
	}
    }
#if PARANOID_ASSERTS
    assert(SetCardinality(outSet) == numOut);
#endif
    unsigned outArray[numOut], Sdeg = 0;
    i = SetToArray(outArray, outSet);
    assert(i == numOut);
    for(i=0; i < numOut; i++)
    {
	switch(connectCount[outArray[i]])
	{
	case 0: break;
	case 1: Sdeg += _k-1; break;
	default:  Sdeg += _k; break;
	}
    }
    SetFree(outSet);
    return Sdeg;
}
#endif

// This function is usually only run at the END after all sampling is finished; it converts graphlet frequencies to
// concentrations or integers based on the sampling algorithm and command line arguments.
// It returns an estimate of the total number of connected graphlets in the entire graph
void convertFrequencies(unsigned long numSamples)
{
    int i;
    assert(numSamples);
    switch(_sampleMethod) {
    case SAMPLE_MCMC: case SAMPLE_NODE_EXPANSION: case SAMPLE_SEQUENTIAL_CHAINING: case SAMPLE_EDGE_EXPANSION:
	if (_freqDisplayMode == freq_display_mode_count || _freqDisplayMode == freq_display_mode_estimate_absolute) {
	    double totalConcentration = 0;
	    for (i = 0; i < _numCanon; i++) assert(_graphletConcentration[i] >= 0.0 && !isinf(_graphletConcentration[i]));
	    for (i = 0; i < _numCanon; i++) totalConcentration += _graphletConcentration[i];
	    assert(totalConcentration);
	    for (i = 0; i < _numCanon; i++) _graphletCount[i] = _graphletConcentration[i]/totalConcentration * numSamples;
	}
	break;
    default:
	if (_freqDisplayMode == freq_display_mode_concentration && numSamples) {
	    for (i = 0; i < _numCanon; i++) {
		assert(_graphletCount[i] >= 0.0 && !isinf(_graphletCount[i]));
		_graphletConcentration[i] = _graphletCount[i] / (double)numSamples;
	    }
	}
	break;
    }
}

double computeAbsoluteMultiplier(unsigned long numSamples)
{
    int i;
    double total=0;
    if (_sampleMethod == SAMPLE_MCMC || _sampleMethod == SAMPLE_NODE_EXPANSION || _sampleMethod == SAMPLE_SEQUENTIAL_CHAINING || _sampleMethod == SAMPLE_EDGE_EXPANSION) {
	long double foundStars = 0;
	for (i = 0; i < _numCanon; i++)
	    if(_canonNumStarMotifs[i] != -1) foundStars += _graphletCount[i]*_canonNumStarMotifs[i];
	if(foundStars) {
	    assert(_totalStarMotifs);
	    _absoluteCountMultiplier = _totalStarMotifs / foundStars;
	    total = _absoluteCountMultiplier * numSamples;
	    //Note("Absolute Count Multiplier %g; estimated total graphlets is %g", _absoluteCountMultiplier, total);
	}
	else
	    if(_totalStarMotifs && _quiet<3) // only print warning if there WERE stars globally but not locally
		Warning("can't estimate absolute count of graphlets because no stars were found among them");
    }
    return total;
}

bool checkDoneSampling(void) {
    if (_stopMode == num_samples) {
        int totalSamples = 0;
        for (int i = 0; i < _numThreads; i++) totalSamples += _threadAccumulators[i].numSamples;
        if (totalSamples >= _numSamples) return true;
    } else if (_stopMode == precision) {
        Apology("Precision stopping not yet supported");
        // Gurjot's code would go here
    } else {
        Apology("No stop mode specified. Specify a number of samples to take (-n) or a precision level to reach (-p) in order to designate how to stop sampling.");
    }
    return false;
}


void* RunBlantInThread(void* arg) {
    // variable unpacking
    ThreadData* args = (ThreadData*)arg;
    // honestly some of this data can just be accessed via global variables
    int k = args->k;
    GRAPH *G = args->G;
    int varraySize = args->varraySize;
    long seed = args->seed;
    int threadId = args->threadId;
    int samplesPerThread = args->samplesPerThread;
    Accumulators *accums = &_threadAccumulators[threadId];
    bool stopThread = false;

    // ODV/GDV vector initializations concern: 
    // you'd be allocating memory for these vectors, multiplied by the number of threads. Is that a lot? Probably slows it down
    // initialize thread local GDV vectors if needed
    if(_outputMode & outputGDV || (_outputMode & communityDetection && _communityMode=='g')) {
        for(int i=0; i<_numCanon; i++) {
            accums->graphletDegreeVector[i] = Ocalloc(G->n, sizeof(**accums->graphletDegreeVector));
            for(int j=0; j<G->n; j++) accums->graphletDegreeVector[i][j]=0.0;
        }
    }
    // initialize thread local ODV vectors if needed
    if(_outputMode & outputODV || (_outputMode & communityDetection && _communityMode=='o')) {
        for(int i=0; i<_numOrbits; i++) {
            accums->orbitDegreeVector[i] = Ocalloc(G->n, sizeof(**accums->orbitDegreeVector));
            for(int j=0; j<G->n; j++) accums->orbitDegreeVector[i][j]=0.0;
        }
    }
    // initialize thread local communityNeighbors if needed
    if(_outputMode & communityDetection) {
        accums->communityNeighbors = (SET***) Calloc(G->n, sizeof(SET**));
    }

    RandomSeed(seed);

#if PARANOID_ASSERTS
    for (int i = 0; i < _numCanon; i++) {
        // printf("graphletConcentration[%d] = %g\n", i, accums->graphletConcentration[i]);
        assert(accums->graphletConcentration[i] == 0);
    }

#endif

    SET *V = SetAlloc(G->n);
    TINY_GRAPH *empty_g = TinyGraphAlloc(k);
    unsigned Varray[varraySize];
    double weight;
    unsigned long stuck = 0;
    SET *prev_node_set = SetAlloc(G->n);
    SET *intersect_node = SetAlloc(G->n);

    if (_outputMode & graphletDistribution) {
        SampleGraphlet(G, V, Varray, k, G->n, &_trashAccumulator);
        SetCopy(prev_node_set, V);
        TinyGraphInducedFromGraph(empty_g, G, Varray);
    }

    int samplesCounter = 0;
    int batchSize = G->numEdges*10;

    // with the index modes, we want specifically the -n number of samples; checkDoneSampling cannot stop after an exact number, so we must handle these seperately
    if ((_outputMode & indexGraphlets || _outputMode & indexGraphletsRNO || _outputMode & indexOrbits) && samplesCounter >= samplesPerThread) stopThread = true;

    // begin sampling
    // printf("Running BLANT in threads to compute %d samples, for k=%d.\n", args->numSamples, k);
    while (!_doneSampling && !stopThread) {
        if (_window) {
            Fatal("Multithreading not yet implemented for any window related output modes.");
        } 
        if (_outputMode & graphletDistribution) {
            // calls SampleGraphlet internally
            ProcessWindowDistribution(G, V, Varray, k, empty_g, prev_node_set, intersect_node);
        } else {
            weight = SampleGraphlet(G, V, Varray, k, G->n, accums);
            if (ProcessGraphlet(G, V, Varray, k, empty_g, weight, accums)) {
                stuck = 0; // reset stuck counter when finding a newly processed graphlet
            } else {
                // processing failed, ignore sample
                samplesCounter--;
                stuck++;
                if(stuck > MAX(G->n, _numSamples)) {
                    if(_quiet<2) Warning("Sampling aborted: no new graphlets discovered after %d attempts", stuck);
                    _doneSampling = true; // no new samples to be found, stop sampling
                }
            }
        }
        samplesCounter++;
        accums->numSamples = samplesCounter; // update the accumulator data, since it's used in global checkDoneSampling()
        // two ways to check if done
        if (_stopMode == precision && samplesCounter && samplesCounter % batchSize == 0 && checkDoneSampling()) {
            _doneSampling = true;
        }
        if (_stopMode == num_samples) {
            if ((_outputMode & indexGraphlets || _outputMode & indexGraphletsRNO || _outputMode & indexOrbits) &&
                samplesCounter >= samplesPerThread) stopThread = true; // stop THIS thread
            else if (checkDoneSampling()) _doneSampling = true; // stop ALL threads
        }
    }
    SetFree(prev_node_set);
    SetFree(intersect_node);
    SetFree(V);
    TinyGraphFree(empty_g);
    pthread_exit(0);
}


// This is the single-threaded BLANT function. YOU PROBABLY SHOULD NOT CALL THIS.
// Call RunBlantInForks (or RunBlantInThread) instead, it's the top-level entry point to call
// once the graph is finished being input---all the ways of reading input call RunBlantIn(Forks|Threads).
// Note it does stuff even if numSamples == 0, because we may be the parent of many
// threads that finished and we have nothing to do except output their accumulated results.
static int RunBlantFromGraph(int k, unsigned long numSamples, GRAPH *G) {
    int windowRepInt, D;
    int i, j;
    unsigned char perm[MAX_K+1];
    assert(k <= G->n);
    assert(k == _k);
    SET *V = SetAlloc(G->n);
    SET *prev_node_set = SetAlloc(G->n);
    SET *intersect_node = SetAlloc(G->n);
#if SELF_LOOPS
    TINY_GRAPH *empty_g = TinyGraphSelfAlloc(k);
#else
    TINY_GRAPH *empty_g = TinyGraphAlloc(k); // allocate it here once, so functions below here don't need to do it repeatedly
#endif
    int varraySize = _windowSize > 0 ? _windowSize : MAX_K + 1;
    unsigned Varray[varraySize];
    InitializeConnectedComponents(G);
    InitializeStarMotifs(G);


    // these initialize the global accumulators, which are the ones being outputed
    // each thread also has it's own copy of accumulators, which are then "summed up" into this global copy we initialize below
    // initialize GDV vectors if needed
    if(_outputMode & outputGDV || (_outputMode & communityDetection && _communityMode=='g')) {
        for(i=0;i<_numCanon;i++) {
            _graphletDegreeVector[i] = Ocalloc(G->n, sizeof(**_graphletDegreeVector));
            for(j=0;j<G->n;j++) _graphletDegreeVector[i][j]=0.0;
        }
    }
    // initialize ODV vectors if needed
    if(_outputMode & outputODV || (_outputMode & communityDetection && _communityMode=='o')) {
        for(i=0;i<_numOrbits;i++) {
            _orbitDegreeVector[i] = Ocalloc(G->n, sizeof(**_orbitDegreeVector));
            for(j=0;j<G->n;j++) _orbitDegreeVector[i][j]=0.0;
        }
    } // note that this double allocation of GDV/ODV vectors may slow things down


    // initialize distribution tables if needed
    if (_outputMode & graphletDistribution) {
        _graphletDistributionTable = Ocalloc(_numCanon, sizeof(int*));
        for(i=0; i<_numCanon; i++) _graphletDistributionTable[i] = Ocalloc(_numCanon, sizeof(int));
        for(i=0; i<_numCanon; i++) for(j=0; j<_numCanon; j++) _graphletDistributionTable[i][j] = 0;
    }

    // initiailize graphlet count/concentrations since they contain garbage
    for (i = 0; i < _numCanon; i++) {
        _graphletConcentration[i] = 0;
        _graphletCount[i] = 0;
    }
    
    if (_sampleMethod == SAMPLE_MCMC)
	_window? initializeMCMC(G, _windowSize, numSamples) : initializeMCMC(G, k, numSamples);
    else if (_sampleMethod == SAMPLE_NODE_EXPANSION || _sampleMethod == SAMPLE_SEQUENTIAL_CHAINING || _sampleMethod == SAMPLE_EDGE_EXPANSION)
	initialize(G, k, numSamples);
    if (_outputMode & graphletDistribution) {
        // accumulators must be provided but no need for them
        SampleGraphlet(G, V, Varray, k, G->n, &_trashAccumulator);
        SetCopy(prev_node_set, V);
        TinyGraphInducedFromGraph(empty_g, G, Varray);
    }

    // if sample method is SAMPLE_INDEX or MCMC, and NOT index output
    if ((_sampleMethod == SAMPLE_INDEX || _sampleSubmethod == SAMPLE_MCMC_EC) &&
	!(_outputMode & indexGraphlets) && !(_outputMode & indexGraphletsRNO) && !(_outputMode & indexOrbits))
	    Fatal("currently only -mi and -mj output modes are supported for INDEX and EDGE_COVER sampling methods");

    // ethan note: nothing written about SAMPLE_INDEX sampling, but it must be single threaded?
    if (_sampleMethod == SAMPLE_INDEX) {
        if (_numThreads > 1) {
            Note("Index sampling must be single threaded, thus only one thread will be ran.");
            _numThreads = 1; // this is unecessary since the below code doesn't use _numThreads at all, but keep anyway
        }

        unsigned prev_nodes_array[_k];

        // Get heuristic values based on orbit number, if ODV file provided
        double heuristicValues[G->n];

        if (_orbitNumber != -1) {
            getOdvValues(heuristicValues, _orbitNumber, _nodeNames, G->n);
        } else {
            getDoubleDegreeArr(heuristicValues, G); // since heuristic values are doubles, we need to convert degree values to doubles
        }

        int percentToPrint = 1;
        node_whn nwhn_arr[G->n]; // nodes sorted first by the heuristic function and then either alphabetically or reverse alphabetically

        // fill node order array with base values
        for (i = 0; i < G->n; i++) {
            nwhn_arr[i].node = i;
            nwhn_arr[i].heur = heuristicValues[i];
            nwhn_arr[i].name = _nodeNames[i];
        }

        // sort array
        int (*comp_func)(const void*, const void*);

        if (_alphabeticTieBreaking) {
            comp_func = nwhn_des_alph_comp_func;
        } else {
            comp_func = nwhn_des_rev_comp_func;
        }

        qsort((void*)nwhn_arr, G->n, sizeof(node_whn), comp_func);

        for(i=0; i<G->n; i++) {
            prev_nodes_array[0] = nwhn_arr[i].node;
            SampleGraphletIndexAndPrint(G, prev_nodes_array, 1, heuristicValues);
            if (i * 100 / G->n >= percentToPrint) {
                fprintf(stderr, "%d%% done\n", percentToPrint);
                ++percentToPrint;
            }
	    }
    }
    else if (_sampleMethod == SAMPLE_MCMC_EC) {
	Fatal("should not get here--EDGE_COVER is a submethod of MCMC");
#if 0
	int e;
	for(e=0; e<G->numEdges; e++) if(SetIn(needEdge, e)) {
	    int whichCC = -(e+1); // encoding edge number in whichCC
	    double weight = SampleGraphlet(G, V, Varray, k, whichCC);
	    ProcessGraphlet(G, V, Varray, k, empty_g, weight);

	    // Now remove all the edges in the graphlet from "needEdge" (SLOW AND DUMB but it's not really a problem)
	    int i,j,f;
	    for(i=0;i<k;i++) for(j=i+1;j<k;j++) {
		int u=Varray[i], v=Varray[j];
		if(GraphAreConnected(G,u,v)) { // find (u,v) in the edgeList and remove it from needEdge
		    Boolean found=false;
		    for(f=0;f<G->numEdges;f++) {
			if((G->edgeList[2*f]==u && G->edgeList[2*f+1]==v) || (G->edgeList[2*f]==v && G->edgeList[2*f+1]==u)) {
			    found=true;
			    SetDelete(needEdge, f);
			    break;
			}
		    }
		    assert(found);
		}
	    }
	}
#endif
    }
    else // sample graphlets from entire graph using either numSamples or confidence
    {
        // Apologize if _numThreads > 1 for a sampling method or output mode that isn't yet supported by multithreading
        if (
            _numThreads > 1 && (
            !(_outputMode & graphletFrequency || _outputMode & outputGDV || _outputMode & outputODV || 
                // index modes have nothing to accumulate, thus work very easily with multithreading
              _outputMode & indexGraphlets || _outputMode & indexGraphletsRNO || _outputMode & indexOrbits ||
              _outputMode & indexMotifs || _outputMode & indexMotifOrbits || _outputMode & communityDetection
            ) ||
            !(_sampleMethod == SAMPLE_EDGE_EXPANSION || _sampleMethod == SAMPLE_NODE_EXPANSION)
            )
        ) {
            Note("Multithreading (t=%d) not supported for the specified output mode or sampling method (%s). Setting number of threads to 1.",  _numThreads, SampleMethodStr());
            _numThreads = 1;
        }

        _threadAccumulators = Ocalloc(_numThreads, sizeof(Accumulators));
        memset(_threadAccumulators, 0, _numThreads * sizeof(Accumulators));
       
        struct timespec start, end;
        clock_gettime(CLOCK_MONOTONIC, &start);

        pthread_t threads[_numThreads];
        ThreadData threadData[_numThreads];
        int samplesPerThread = numSamples / _numThreads;
        int leftover = numSamples - (samplesPerThread * _numThreads);
        // seed the threads with a base seed that may or may not be specified
        long base_seed = _seed == -1 ? GetFancySeed(true) : _seed;

        Note("Running BLANT in %d threads", _numThreads);

        for (unsigned t = 0; t < _numThreads; t++)
        {
            threadData[t].samplesPerThread = samplesPerThread;
            if (t == _numThreads - 1) threadData[t].samplesPerThread += (leftover);
            threadData[t].k = k;
            threadData[t].G = G;
            threadData[t].varraySize = varraySize;
            threadData[t].threadId = t;
            threadData[t].seed = base_seed + t; // each thread has it's own unique seed
            pthread_create(&threads[t], NULL, RunBlantInThread, &threadData[t]);
        }

        // join threads and then summate the accumulators from each thread
        for (unsigned t = 0; t < _numThreads; t++) {
            pthread_join(threads[t], NULL);
        }

        for (unsigned t = 0; t < _numThreads; t++) {
            for (i = 0; i < _numCanon; i++) {
                _graphletConcentration[i] += _threadAccumulators[t].graphletConcentration[i];
                _graphletCount[i] += _threadAccumulators[t].graphletCount[i];
            }

            if (_outputMode & outputODV) {
                for(i=0; i<_numOrbits; i++) {
                for(j=0; j<G->n; j++) {
                    // if (t == 0) assert(_orbitDegreeVector[i][j] == 0);
                    _orbitDegreeVector[i][j] += _threadAccumulators[t].orbitDegreeVector[i][j];
                }
                }
            }
            if (_outputMode & outputGDV) {
                for(i=0; i<_numCanon; i++) {
                for(j=0; j<G->n; j++) {
                    // if (t == 0) assert(_graphletDegreeVector[i][j] == 0);
                    // if (t == 0 &&_graphletDegreeVector[i][j] != 0) {
                    //     Note("BAD VALUE: %f", _graphletDegreeVector[i][j]);
                    //     assert(false);
                    // }
                    _graphletDegreeVector[i][j] += _threadAccumulators[t].graphletDegreeVector[i][j];
                }
                }
            }
            // Consolidate community neighbors
            if (_outputMode & communityDetection) {
				switch(_communityMode) {
					case 'o':
						for(i=0; i<_numOrbits; i++) {
							for(j=0; j<G->n; j++) {
									_orbitDegreeVector[i][j] += _threadAccumulators->orbitDegreeVector[i][j];
							}
						}
						break;
					case 'g':
						for(i=0; i<_numCanon; i++) {
							for(j=0; j<G->n; j++) {
									_graphletDegreeVector[i][j] += _threadAccumulators->graphletDegreeVector[i][j];
							}
						}
						break;
				}
				int numCommunities = (_communityMode=='o') ? _numOrbits : _numCanon;
				for(i=0; i<G->n; i++) {
					if(_threadAccumulators->communityNeighbors[i]) {
						if(!_communityNeighbors[i]) {
							_communityNeighbors[i] = (SET**) Calloc(numCommunities, sizeof(SET*));
						}
						for(j=0; j<numCommunities; j++) {
							if(_threadAccumulators->communityNeighbors[i][j]) {
								if(!_communityNeighbors[i][j]) {
									_communityNeighbors[i][j] = SetAlloc(G->n);
								}
								_communityNeighbors[i][j] = SetUnion(_communityNeighbors[i][j], _communityNeighbors[i][j], _threadAccumulators->communityNeighbors[i][j]);
							}
						}
					}
				}
            }
        }

        clock_gettime(CLOCK_MONOTONIC, &end);
        double elapsed_time = (end.tv_sec - start.tv_sec) +
                            (end.tv_nsec - start.tv_nsec) / 1e9;
        Note("Took %f seconds to sample %d with %d threads.", elapsed_time, numSamples, _numThreads); 

    #if 0
	int batchSize = G->numEdges*10; // 300000; //1000*sqrt(_numOrbits); //heuristic: batchSizes smaller than this lead to spurious early stops
	if(_desiredPrec && _quiet<2)
	    Note("using batches of size %d to estimate counts with relative precision %g (%g digit%s) with %g%% confidence",
		batchSize, _desiredPrec, _desiredDigits, (fabs(1-_desiredDigits)<1e-6?"":"s"), 100*_confidence);
	unsigned long i;
	STAT *sTotal[MAX_CANONICALS];
	for(i=0; i<_numCanon; i++) if(SetIn(_connectedCanonicals,i)) sTotal[i] = StatAlloc(0,0,0, false, false);
	Boolean confMet = false;
	static int batch;
        for(i=0; (i<numSamples || (_sampleFile && !_sampleFileEOF) || (_desiredPrec && !confMet)) && !_earlyAbort; i++)
        {
            if(_window) {
                SampleGraphlet(G, V, Varray, _windowSize, G->n);
                _numWindowRep = 0;
                if (_windowSampleMethod == WINDOW_SAMPLE_MIN || _windowSampleMethod == WINDOW_SAMPLE_MIN_D ||
			_windowSampleMethod == WINDOW_SAMPLE_LEAST_FREQ_MIN)
                    windowRepInt = maxBk;
                if (_windowSampleMethod == WINDOW_SAMPLE_MAX || _windowSampleMethod == WINDOW_SAMPLE_MAX_D ||
			_windowSampleMethod == WINDOW_SAMPLE_LEAST_FREQ_MAX)
                    windowRepInt = -1;
                D = _k * (_k - 1) / 2;
                FindWindowRepInWindow(G, V, &windowRepInt, &D, perm);
                if(_numWindowRep > 0)
                    ProcessWindowRep(G, Varray, windowRepInt);
            }
            else if (_outputMode & graphletDistribution)
                ProcessWindowDistribution(G, V, Varray, k, empty_g, prev_node_set, intersect_node);
            else {
		static unsigned long stuck;
                double weight = SampleGraphlet(G, V, Varray, k, G->n);
                if(ProcessGraphlet(G, V, Varray, k, empty_g, weight)) {
		    stuck = 0;
		    if(_desiredPrec) {
			if(i && i%batchSize==0) { // we just finished a batch
			    int minNumBatches = 13-k+1/sqrt(1-_confidence)/k; //heuristic
			    int maxNumBatches = 1000*minNumBatches; // huge
			    double worstInterval=0, intervalSum=0;
			    _worstCanon = -1;
			    if(_batchRawTotalSamples) { // in rare cases a batch may finish with no actual samples
				int j;
				// Even though the samples may not be Normally distributed, the Law of Large Numbers
				// guarantees that for sufficiently large batches, the batch means *are* Normally
				// distributed, so we can compute confidence intervals.
				for(j=0;j<_numCanon;j++) if(SetIn(_connectedCanonicals,j) && _batchRawCount[j]) {
				    double sample = _batchRawCount[j];
				    if(_precisionWt == PrecWtNone) StatAddSample(sTotal[j], sample);
				    else {
					double w = sample;
					if(_precisionWt == PrecWtLog) {if(w>1) w = log(w);}
					else assert(_precisionWt == PrecWtRaw);
					StatAddWeightedSample(sTotal[j], w, sample);
				    }
				    if(StatN(sTotal[j]) > 1) {
					double relInterval = StatConfInterval(sTotal[j], _confidence) / StatMean(sTotal[j]);
					intervalSum += relInterval;
					// under-sampled graphlets (<=2) don't count towards "worst"
					if( _graphletCount[j] > 2 && relInterval > worstInterval) {
					    worstInterval = relInterval;
					    _worstCanon=j;
					}
				    }
				}
				_meanPrec = intervalSum/_numConnectedCanon;
				if(worstInterval) _worstPrecision = worstInterval;
				double precision=0;
				switch(_precisionMode) {
				    case mean: precision = _meanPrec; break;
				    case worst: precision = _worstPrecision; break;
				    default: Fatal("unknown precision mode"); break;
				}

				if(++batch && _quiet<1) {
				    FILE *fp;
				    fp = popen("date -Iseconds | sed -e 's/T/ /' -e 's/,/./' -e 's/-..:..$//'", "r");
				    char buf[BUFSIZ];
				    if(fp && buf == fgets(buf, sizeof(buf)-1, fp)) {
					pclose(fp);
					buf[strlen(buf)-1] = '\0'; // nuke the newline
				    }
				    else strcpy(buf, "(time failed)");
				    Note("%s batch %d CPU %gs samples %ld prec mean %.3g worst %.3g (g%d count %.0f)",buf,batch,
					GetCPUseconds(), i, _meanPrec, worstInterval, _worstCanon, _graphletCount[_worstCanon]);
				}
				if(batch>=maxNumBatches || (batch >= minNumBatches && precision < _desiredPrec))
				    confMet=true; // don't reset the counts if we're done
				else {
				    _batchRawTotalSamples = 0;
				    for(j=0;j<_numCanon;j++) _batchRawCount[j] = 0;
				}
			    }
			    else
				if(_quiet<3) Warning("invalid batch %d, batchTotal is zero", ++batch);
			}
		    }
		}
		else { // Processing failed--likely a recent duplicate detected by NodeSetSeenRecently().
		    if(numSamples) --i; // negate the sample count of duplicate graphlets
		    ++stuck;
		    if(stuck > MAX(G->n,numSamples)) {
			if(_quiet<2) Warning("Sampling aborted: no new graphlets discovered after %d attempts", stuck);
			_earlyAbort = true;
		    }
		}
            }
        }
	// assert(i);
	if(i<numSamples) {
	    if(_quiet<2) Warning("only took %d samples out of %d", i, numSamples);
	}
	else
	{
	    if((_sampleFile && _sampleFileEOF) || (_desiredPrec && confMet) || _earlyAbort) {
		_numSamples = numSamples = i-1; // lots of code below assumes numSamples is set later on
		if(!_quiet) Note("numSamples was %lu", numSamples);
	    }
	    if(_desiredPrec && confMet && _quiet<2)
		Note("estimated precision %.1f digits (worst %.1f ID %d) at %g%% confidence after %lu samples in %d batches",
		    -(log(_meanPrec?_meanPrec:1)/log(10)), -(log(_worstPrecision?_worstPrecision:1)/log(10)), _worstCanon,
		    100*_confidence, numSamples, batch);
	}
	for(i=0; i<_numCanon; i++) if(SetIn(_connectedCanonicals,i)) StatFree(sTotal[i]);
    #endif
    }

    // Sampling done. Now generate output for output modes that require it.

    if(_window) {
        for(i=0; i<_numWindowRepArrSize; i++) Free(_windowReps[i]);
        Free(_windowReps);
        if(_windowRep_limit_method) HeapFree(_windowRep_limit_heap);
    }
    if ((_sampleMethod==SAMPLE_MCMC || _sampleMethod==SAMPLE_NODE_EXPANSION || _sampleMethod==SAMPLE_SEQUENTIAL_CHAINING || _sampleMethod == SAMPLE_EDGE_EXPANSION) &&
	    !_window)
	finalize(G, numSamples);

    if ((_outputMode & graphletFrequency || _outputMode & outputGDV || _outputMode & outputODV) && !_window)
	convertFrequencies(numSamples);

    if(((_outputMode & graphletFrequency && _freqDisplayMode == freq_display_mode_estimate_absolute) ||
	_outputMode & outputGDV || _outputMode & outputODV) && !_window)
	computeAbsoluteMultiplier(numSamples);

    int canon, orbit_index, u,v,c;
    // case indexGraphlets: case indexGraphletsRNO: case indexOrbits: case indexMotifs: case indexMotifOrbits:
	//break; // already printed on-the-fly in the Sample/Process loop above
    if(_outputMode & graphletFrequency) {
	char buf[BUFSIZ];
	for(canon=0; canon<_numCanon; canon++) {
	    if (SetIn(_connectedCanonicals, canon)) {
		if(_freqDisplayMode == freq_display_mode_concentration)
		    printf("%lf %s\n", _graphletConcentration[canon], PrintOrdinal(buf, canon));
		else
		    printf("%llu %s\n", (unsigned long long) llround(_absoluteCountMultiplier * _graphletCount[canon]),
			PrintOrdinal(buf, canon));
	    }
	}
    }
    if(_outputMode & predict_merge) assert(false); // shouldn't get here
    if(_outputMode & predict) {
	Predict_Flush(G);
	// Turn off ODV and GDV output because we needed them to be COMPUTED but don't want them output in predictMode
	_outputMode &= ~(outputGDV|outputODV);
    }
    // Note: the test for predict mode output MUST be above the test for ODV/GDV output, to turn them off if we're predicting
    if(_outputMode & outputGDV) {
	for(i=0; i < G->n; i++)
	{
	    char buf[BUFSIZ];
	    PrintNode(buf, 0,i);
	    for(canon=0; canon < _numCanon; canon++)
		if (_MCMC_EVERY_EDGE || (_sampleMethod != SAMPLE_MCMC && _sampleMethod != SAMPLE_NODE_EXPANSION && 
			_sampleMethod != SAMPLE_SEQUENTIAL_CHAINING && _sampleMethod != SAMPLE_EDGE_EXPANSION)) 
            sprintf(buf+strlen(buf), " %.15g", GDV(i,canon));
		else sprintf(buf+strlen(buf), " %llu", (unsigned long long) llround(_absoluteCountMultiplier * _graphletDegreeVector[canon][i]));
	    assert(strlen(buf) < BUFSIZ);
	    puts(buf);
	}
    }
    if(_outputMode & outputODV) {
	    char buf[BUFSIZ];
        for(i=0; i<G->n; i++) {
	    PrintNode(buf,0,i);
	    for(j=0; j<_numConnectedOrbits; j++) {
		if (k == 4 || k == 5) orbit_index = _connectedOrbits[_orca_orbit_mapping[j]];
		else orbit_index = _connectedOrbits[j];
		if(_MCMC_EVERY_EDGE || (_sampleMethod != SAMPLE_MCMC && _sampleMethod != SAMPLE_NODE_EXPANSION && 
		    _sampleMethod != SAMPLE_SEQUENTIAL_CHAINING && _sampleMethod != SAMPLE_EDGE_EXPANSION))
			sprintf(buf+strlen(buf), " %.15g", ODV(i,orbit_index));
		else sprintf(buf+strlen(buf), " %llu", (unsigned long long) llround(_absoluteCountMultiplier * _orbitDegreeVector[orbit_index][i]));
	    }
	    assert(strlen(buf) < BUFSIZ);
	    puts(buf);
	}
    }
    if(_outputMode & communityDetection) {
	assert(_communityMode == 'o' || _communityMode == 'g');
	switch(_communityMode)
	{
	case 'o':
	    for(u=0; u<G->n; u++) {
		for(c=0; c<_numConnectedOrbits; c++) {
		    orbit_index = _connectedOrbits[c];
		    double odv=_orbitDegreeVector[orbit_index][u];
		    int numPrinted = 0;
		    char buf[BUFSIZ];
		    if(odv && _communityNeighbors[u] && _communityNeighbors[u][orbit_index] && SetCardinality(_communityNeighbors[u][orbit_index])) {
			int neigh=0;
			for(j=0;j<GraphDegree(G,u); j++) {
			    char jbuf[BUFSIZ];
			    v = GraphNextNeighbor(G,u,&neigh); assert(v!=-1); // G->neighbor[u][j];
			    if(SetIn(_communityNeighbors[u][orbit_index],v)) {
				if(!numPrinted++) sprintf(buf, "%s %d %.15g\t", PrintNode(jbuf, 0,u), orbit_index, odv);
				PrintNode(buf+strlen(buf), ' ',v);
			    }
			}
			assert(-1==GraphNextNeighbor(G,u,&neigh));
			assert(strlen(buf) < BUFSIZ);
			if(numPrinted) puts(buf);
		    }
		}
	    }
	    break;
	case 'g':
	    for(u=0; u<G->n; u++) {
		for(c=0; c<_numCanon; c++) if(SetIn(_connectedCanonicals,c)) {
		    double gdv=_graphletDegreeVector[c][u];
		    int numPrinted = 0;
		    char buf[BUFSIZ];
		    if(gdv && _communityNeighbors[u] && _communityNeighbors[u][c] && SetCardinality(_communityNeighbors[u][c])) {
			int neigh=0;
			for(j=0;j<GraphDegree(G,u); j++) {
			    char jbuf[BUFSIZ];
			    v = GraphNextNeighbor(G,u,&neigh); assert(v!=-1); //G->neighbor[u][j];
			    if(SetIn(_communityNeighbors[u][c],v)) {
				if(!numPrinted++) sprintf(buf, "%s %d %lg\t", PrintNode(jbuf, 0,u), c, gdv);
				PrintNode(buf+strlen(buf), ' ',v);
			    }
			}
			assert(-1==GraphNextNeighbor(G,u,&neigh));
			assert(strlen(buf) < BUFSIZ);
			if(numPrinted) puts(buf);
		    }
		}
	    }
	    break;
	default: Fatal("unknown _communityMode %c", _communityMode);
	}
    }
    if(_outputMode & graphletDistribution) {
        for(i=0; i<_numCanon; i++) {
            for(j=0; j<_numCanon; j++)
                printf("%d ", _graphletDistributionTable[i][j]);
            printf("\n");
        }
    }
    if(!_outputMode) Abort("RunBlantFromGraph: unknown or un-implemented outputMode");

#if !O_ALLOC && PARANOID_ASSERTS
    // no point in freeing this stuff since we're about to exit; it can take significant time for large graphs.
    if(_outputMode & outputGDV) for(i=0;i<_numCanon;i++) Free(_graphletDegreeVector[i]);
    if(_outputMode & outputODV || (_outputMode & communityDetection && _communityMode=='o'))
	for(i=0;i<_numOrbits;i++) Free(_orbitDegreeVector[i]);
    if(_outputMode & outputODV && _MCMC_EVERY_EDGE) for(i=0;i<_numOrbits;i++) Free(_orbitDegreeVector[i]);
#endif
    if (_sampleMethod == SAMPLE_ACCEPT_REJECT && numSamples)
    	fprintf(stderr,"Average number of tries per sample is %.15g\n", _acceptRejectTotalTries/(double)numSamples);
    SetFree(V);
    SetFree(prev_node_set);
    SetFree(intersect_node);
    TinyGraphFree(empty_g);
    return _earlyAbort;
}

/*
** Fork a BLANT process and return a FILE pointer where it'll be sending stuff.
** Caller is responsible for reading all the stuff from the returned FILE pointer,
** detecting EOF on it, and fclose'ing it.
*/
FILE *ForkBlant(int k, unsigned long numSamples, GRAPH *G)
{
    int fds[2] = {-1, -1};
    if(pipe(fds) < 0) Fatal("pipe(2) failed");
    int threadSeed = INT_MAX*RandomUniform(); // this must happen BEFORE the fork for each thread to get a different seed
    int pid = fork();
    if(pid > 0) // we are the parent
    {
	close(fds[1]); // we will not be writing to the pipe, so close it.
	return fdopen(fds[0],"r");
    }
    else if(pid == 0) // we are the child
    {
	_child = true;
	_seed = threadSeed;
	RandomSeed(_seed);
	(void)close(fds[0]); // we will not be reading from the pipe, so close it.
	(void)close(1); // close our usual stdout
	if(dup(fds[1])<0) Fatal("dup(2) failed"); // copy the write end of the pipe to fd 1.
	(void)close(fds[1]); // close the original write end of the pipe since it's been moved to fd 1.

	// For any "counting" mode, use internal numbering when communicating through pipes to the parent
	if(!(_outputMode & indexGraphlets) && !(_outputMode & indexGraphletsRNO) && !(_outputMode & indexOrbits))
	    _supportNodeNames = false;

	RunBlantFromGraph(k, numSamples, G);
	exit(0);
	_exit(0);
	Abort("Both exit() and _exit failed???");
    }
    else
	Abort("fork failed");
    assert(false);
    return NULL; //  should never get here, but quell compiler warning.
}


static FILE *fpProcs[MAX_POSSIBLE_THREADS]; // these will be the pipes reading output of the parallel blants

// This is the primary entry point into BLANT, even if THREADS=1.  We assume you've already
// read the graph into G, and will do whatever is necessary to run blant with the number of
// child processes specified.  Also does some sanity checking. 
// [PTHREADS BRANCH IS DEPRECATING THIS]
int RunBlantInForks(int k, unsigned long numSamples, GRAPH *G)
{
    int i, j;
    assert(k == _k);
    assert(G->n >= k); // should really ensure at least one connected component has >=k nodes. TODO
    if(_outputMode & outputGDV || (_outputMode & communityDetection && _communityMode=='g')) {
	// for(i=0;i<_numCanon;i++) _graphletDegreeVector[i] = Ocalloc(G->n, sizeof(**_graphletDegreeVector));
	for(i=0;i<_numCanon;i++) _graphletDegreeVector[i] = Ocalloc(G->n, sizeof(**_graphletDegreeVector));
    }
    if(_outputMode & outputODV || (_outputMode & communityDetection && _communityMode=='o')) {
	for(i=0;i<_numOrbits;i++) {
	    // _orbitDegreeVector[i] = Ocalloc(G->n, sizeof(**_orbitDegreeVector));
	    // for(j=0;j<G->n;j++) _orbitDegreeVector[i][j]=0;
	}
    }
    if (_outputMode & outputODV) for(i=0;i<_numOrbits;i++){
	_orbitDegreeVector[i] = Ocalloc(G->n, sizeof(**_orbitDegreeVector));
	for(j=0;j<G->n;j++) _orbitDegreeVector[i][j]=0.0;
    }
    if(_outputMode & predict) Predict_Init(G);
    if (_outputMode & graphletDistribution) {
        _graphletDistributionTable = Ocalloc(_numCanon, sizeof(int*));
        for(i=0; i<_numCanon; i++) _graphletDistributionTable[i] = Ocalloc(_numCanon, sizeof(int));
        for(i=0; i<_numCanon; i++) for(j=0; j<_numCanon; j++) _graphletDistributionTable[i][j] = 0;
    }

    if(_numThreads == 1)
	return RunBlantFromGraph(k, numSamples, G);

    if (_sampleMethod == SAMPLE_INDEX || _sampleSubmethod == SAMPLE_MCMC_EC)
        Fatal("The sampling methods INDEX and EDGE_COVER do not yet support multithreading (feel free to add it!)");

    // At this point, _JOBS must be greater than 1.
    assert(_numThreads>1 && _numThreads <= _maxThreads);
    unsigned long totalSamples = numSamples;
    double meanSamplesPerJob = totalSamples/(double)_numThreads;
    if(!_quiet) Note("Parent %d starting about %d jobs of about %d samples each", getpid(), _numThreads, (int)meanSamplesPerJob);

    int procsRunning = 0, jobsDone = 0, proc, job=0;
    unsigned long lineNum = 0;
    for(i=0; numSamples > 0 && i<_maxThreads;i++) {
	unsigned long samples = meanSamplesPerJob;
	assert(samples>0);
	if(samples > numSamples) samples = numSamples;
	numSamples -= samples;
	fpProcs[i] = ForkBlant(_k, samples, G);
	if(!_quiet) Note("Started job %d of %d samples; %d children running, %ld samples remaining to take",
	    job++, samples, ++procsRunning, numSamples);
    }

    do // this loop reads lines from the parallel child procs, one line read per proc per loop iteration
    {
	char line[1000 * BUFSIZ]; // this is just supposed to be really large
	for(proc=0;proc<_maxThreads;proc++)	// process one line from each proc
	{
	    if(!fpProcs[proc]) continue;
	    char *tmp = fgets(line, sizeof(line), fpProcs[proc]);
	    assert(tmp>=0);
	    if(feof(fpProcs[proc]))
	    {
		wait(NULL); // reap the child
		clearerr(fpProcs[proc]);
		fclose(fpProcs[proc]);
		fpProcs[proc] = NULL;
		++jobsDone; --procsRunning;
		if(!_quiet) Note("Thead %d finished; jobsDone %d, procsRunning %d", proc, jobsDone, procsRunning);
		if(numSamples == 0) fpProcs[proc] = NULL; // signify this pointer is finished.
		else {
		    unsigned long samples = meanSamplesPerJob;
		    if(samples > numSamples) samples = numSamples;
		    numSamples -= samples;
		    fpProcs[proc] = ForkBlant(_k, samples, G);
		    assert(fpProcs[proc]);
		    ++procsRunning;
		    if(!_quiet)
			Note("Started job %d (proc %d) of %d samples, %d procs running, %ld samples remaining to take",
			    job++, proc, samples, procsRunning, numSamples);
		}
		continue; // we'll ask for output next time around.
	    }
	    char *nextChar = line, *pch;
	    double gcount = 0.0;
	    // EDWARD: printf("Line %d from proc %d is \"%s\"\n", lineNum, proc, tmp);
	    int canon=-1, orbit=-1, numRead, nodeId;
	    //fprintf(stderr, "Parent received the following line from the child: <%s>\n", line);
	    if(_outputMode & graphletFrequency) {
		numRead = scanf(line, "%lf%d", &gcount, &canon);
		assert(numRead == 2);
		_graphletCount[canon] += gcount;
	    }
	    if(_outputMode & graphletDistribution) {
		for(i=0; i<_numCanon; i++) {
		    pch = strtok(line, " ");
		    for(j=0; j<_numCanon; j++){
			_graphletDistributionTable[i][j] += atoi(pch);
			pch = strtok(NULL, " ");
		    }
		    char *OK = fgets(line, sizeof(line), fpProcs[proc]);
		    assert(OK || (i==_numCanon-1 && j == _numCanon));
		}
	    }
	    if(_outputMode & outputGDV) {
		assert(isdigit(*nextChar));
		numRead = sscanf(nextChar, "%d", &nodeId);
		assert(numRead == 1 && nodeId == lineNum);
		while(isdigit(*nextChar)) nextChar++; // read past current integer
		assert(*nextChar == ' ' || (canon == _numCanon-1 && *nextChar == '\n'));
		nextChar++;
		for(canon=0; canon < _numCanon; canon++)
		{
		    assert(isdigit(*nextChar));
		    numRead = sscanf(nextChar, "%lf", &gcount);
		    assert(numRead == 1);
		    GDV(lineNum,canon) += gcount;
		    while(isdigit(*nextChar)) nextChar++; // read past current integer
		    assert(*nextChar == ' ' || (canon == _numCanon-1 && *nextChar == '\n'));
		    nextChar++;
		}
		assert(*nextChar == '\0');
	    }
	    if(_outputMode & outputODV) {
		assert(isdigit(*nextChar));
		numRead = sscanf(nextChar, "%d", &nodeId);
		assert(numRead == 1 && nodeId == lineNum);
		while(isdigit(*nextChar)) nextChar++; // read past current integer
		assert(*nextChar == ' ' || (orbit == _numOrbits-1 && *nextChar == '\n'));
		nextChar++;
		for(orbit=0; orbit < _numOrbits; orbit++)
		{
		    assert(isdigit(*nextChar));
		    numRead = sscanf(nextChar, "%lf", &gcount);
		    assert(numRead == 1);
		    ODV(lineNum,orbit) += gcount;
		    while(isdigit(*nextChar)) nextChar++; // read past current integer
		    assert(*nextChar == ' ' || (orbit == _numOrbits-1 && *nextChar == '\n'));
		    nextChar++;
		}
		assert(*nextChar == '\0');
	    }
	    if(_outputMode&indexGraphlets || _outputMode&indexGraphletsRNO || _outputMode&indexOrbits || _outputMode&indexMotifs || _outputMode&indexMotifOrbits) {
		fputs(line, stdout);
		if(_window)
		    while(fgets(line, sizeof(line), fpProcs[proc]))
			fputs(line, stdout);
	    }
	    if(_outputMode&predict_merge) assert(false); // should not be here
	    if(_outputMode& predict) Predict_ProcessLine(G, lineNum, line);
	    if(!_outputMode)
		Abort("oops... unknown or unsupported _outputMode in RunBlantInForks while reading child process");
	}
	lineNum++;
	if(procsRunning > 0) Fatal("Need a way to pass StarMotifCounts from children");
    } while(procsRunning > 0);

    // if numSamples is not a multiple of _THREADS, finish the leftover samples
    unsigned long leftovers = numSamples % _numThreads;
    int result = RunBlantFromGraph(_k, leftovers, G);
    if(_outputMode & predict) Predict_Shutdown(G);
    return result;
}

void BlantAddEdge(int v1, int v2, double weight)
{
    if(!_pairs) _pairs = Malloc(2*_maxEdges*sizeof(_pairs[0]));
    if(_weighted) {assert(weight); if(!_weights) _weights = Malloc(_maxEdges*sizeof(_weights[0]));}
    else assert(weight==0.0);
    assert(_numEdges <= _maxEdges);
    if(_numEdges >= _maxEdges)
    {
	_maxEdges *=2;
	_pairs = Realloc(_pairs, 2*_maxEdges*sizeof(_pairs[0]));
	if(_weighted) _weights=Realloc(_weights, _maxEdges*sizeof(_weights[0]));
    }
    _numNodes = MAX(_numNodes, v1+1); // add one since, for example, if we see a node numbered 100, numNodes is 101.
    _numNodes = MAX(_numNodes, v2+1);
    _pairs[2*_numEdges] = v1;
    _pairs[2*_numEdges+1] = v2;
    if(_weighted) _weights[_numEdges] = weight;
    if(_pairs[2*_numEdges] == _pairs[2*_numEdges+1])
	Fatal("BlantAddEdge: edge %d (%d,%d) has equal nodes; cannot have self-loops\n", _numEdges, v1, v2);
    if(_pairs[2*_numEdges] > _pairs[2*_numEdges+1])
    {
	int tmp = _pairs[2*_numEdges];
	_pairs[2*_numEdges] = _pairs[2*_numEdges+1];
	_pairs[2*_numEdges+1] = tmp;
    }
    assert(_pairs[2*_numEdges] < _pairs[2*_numEdges+1]);
    _numEdges++;
}

int RunBlantEdgesFinished(int k, unsigned long numSamples, int numNodes, char **nodeNames)
{
    GRAPH *G = GraphFromEdgeList(_numNodes, _numEdges, _pairs, SPARSE, _weights);
    Free(_pairs);
    _nodeNames = nodeNames;
    return RunBlantInForks(k, numSamples, G);
}

// Initialize the graph G from an edgelist; the user must allocate the pairs array
// to have 2*numEdges elements (all integers), and each entry must be between 0 and
// numNodes-1. The pairs array MUST be allocated using malloc or calloc, because
// we are going to free it right after creating G (ie., before returning to the caller.)
int RunBlantFromEdgeList(int k, unsigned long numSamples, int numNodes, int numEdges, unsigned *pairs, float *weights)
{
    assert(numNodes >= k);
    GRAPH *G = GraphFromEdgeList(numNodes, numEdges, pairs, SPARSE, weights);
    Free(pairs);
    return RunBlantInForks(k, numSamples, G);
}

const char * const USAGE_SHORT =
"BLANT (Basic Local Alignment of Network Topology): sample graphlets of up to 8 nodes from a graph.\n"\
"USAGE: blant [OPTIONS] -k graphletNodes graphInputFile\n"\
" Common options: (use -h for longer help)\n"\
"    -s samplingMethod (default MCMC; SEC, NBE, EBE!, RES!, AR!, FAYE!, INDEX, EDGE_COVER)\n"\
"       Note: exclamation mark required after some method names to supress warnings about them.\n"\
"    -m{outputMode} (default f=frequency; o=ODV, g=GDV, i=index, cX=community(X=g,o), r=root, d=neighbor distribution\n"\
"    -d{displayModeForCanonicalIDs} (default i=integerOrdinal, o=ORCA, j=Jesse, b=binary, d=decimal, n=noncanonical)\n"\
"    -r seed (integer)\n"\
"    -F frequency display mode (default a=absolute count (estimated); c=concentration; n=raw count out of numSamples)\n"\
"    -t N[:M]: (CURRENTLY BROKEN): use threading (parallelism); break the task up into N jobs (default 1) allowing\n"\
"        at most M to run at one time; M can be anything from 1 to a compile-time-specified maximum possible value\n"\
"        (MAX_POSSIBLE_THREADS in blant.h), but defaults to 4 to be conservative.";

const char * const USAGE_LONG =
"BLANT: Basic Local Alignment of Network Topology (work in progress)\n"\
"PURPOSE: sample graphlets of up to 8 nodes from a graph. Default output is similar to ORCA, though via stochastic sampling\n"\
"    rather than exaustive enumeration. Our APPROXIMATE results come MUCH faster than ORCA on large or dense networks.\n"\
"USAGE: blant [OPTIONS] -k numNodes {-[pP] precision | -n numSamples} graphInputFile\n"\
"where the following are REQUIRED:\n"\
"    numNodes is an integer 3 through 8 inclusive, specifying the size (in nodes) of graphlets to sample;\n"\
"    precision: desired precision of graphlet frequencies; use p<1 for fractional precision, p>=1 for #digits (base 10).\n"\
"        Note0: when p>=1, digits can be non-integer, eg 1.5 means fractional precision 10^(-1.5) or about 3% precision.\n"
"        Note1: -p limits the MEAN precision across graphlets; -P limits the worst case precision (not recommended)\n"\
"        Note2: if a 'w' or 'W' is appended after the number specifying precision, then the mean will be WEIGHTED\n"\
"               by the count--effectively saying you don't care about rare graphlets and want the frequent ones to have\n"\
"               accurate counts. If an 'l' or 'L' is appended, we weigh by the logarithm of the count, which is less\n"\
"               Draconian against rare graphlets.\n"\
"        Note3: technically we use confidence intervals. The relative interval width is set to the precision,\n"\
"               and the default confidence is (1-precision/10). eg. if p=0.01 (2 digits), confidence is set to 99.9%;\n"\
"               p=0.001 (3 digits) sets confidence to 99.99%. The confidence is applied to the mean if -p was specified,\n"\
"               or the worst-case if -P was specified (not recommended). To change the default confidence, use the\n"\
"               -c option (confidence >=1 is assumed to mean percentage\n"\
"    numSamples is the number of graphlet samples to take (large samples are recommended), except in INDEX sampling mode,\n"\
"	where it specifies the maximum number of samples to take from each node in the graph.\n"\
"       Note: this option is mutually exclusive to -p or -P.\n"\
"    samplingMethod is:\n"\
"	MCMC (Markov Chain Monte Carlo): Build the first set S of k nodes using NBE; then randomly remove and add\n"\
"         nodes to S using an MCMC graph walking algorithm with restarts; gives asymptotically correct relative frequencies\n"\
"         when using purely counting modes like -m{o|g|f}, but biased counts in indexing modes like -m{i|j} since we remove\n"\
"         duplicates in indexing modes.)\n"\
"	SEC (Sequential Edge Chaining): like MCMC but instead of a walk, reset the walk for every sampled graphlet.\n"\
"         Note: SEC is the default when using precision (-[Pp] on the command line), otherwise MCMC is the default.\n"\
"	NBE (Node-Based Expansion): pick an edge at random and add it to the node set S; add new nodes by choosing\n"\
"         uniformly at random from all nodes one step outside S. (slow, but correct)\n"\
"	EBE (Edge-Based Expansion): pick an edge at random and add its two nodes to S; add nodes to S by picking an edge\n"\
"         uniformly at random from those emanating from S. (faster than NBE on dense networks, but more biased)\n"\
"	RES (Lu Bressan's REServoir sampling): also asymptotically correct but much slower than MCMC.\n"\
"	AR (Accept-Reject): EXTREMELY SLOW but asymptotically correct: pick k nodes entirely at random, reject if\n"\
"	  resulting graphlet is disconnected (vast majority of such grpahlets are disconnected, thus VERY SLOW)\n"\
"	INDEX: deterministic: for each node v in the graph, build a topologically deterministic set of k-graphlets to\n"\
"         be used as indices for seed-and-extend local alignments (using, eg., our onw Dijkstra-inspired local aligner--\n"\
"         see Dijkstra diretory). When using INDEX sampling, the -n command-line option specifies the maximum number\n"\
"         of index entries per starting node v.\n"\
"	EDGE_COVER: starting with E=all edges in the input graph, pick one edge in E and build a k-graphlet using EBE;\n"\
"         then subtract ALL its edges from E. Continue until E is empty. The goal is to output a short list of graphlets\n"\
"         that, in comglomerate, cover each edge in E at least once.\n"\
"    graphInputFile: graph must be in one of the following formats with its extension name:\n"\
"	Edgelist (.el), LEDA(.leda), GML (.gml), GraphML (.xml), LGF(.lgf), CSV(.csv)\n"\
"	(extensions .gz and .xz are automatically decompressed using gunzip and unxz, respectively)\n"\
"	Duplicate edges (either direction) and self-loops should be removed!\n"\
"COMMON OPTIONS:\n"\
"    -m{outputMode}, where {outputMode} is a single character, one of:\n"\
"	f = [default] graphlet {f}requency, estimates the total count of each type of graphlet.\n"\
"	    sub-option -mf{freqDispMode} can be e{default=estimate), n(integer count) or c(concentration)\n"\
"	o = ODV (Orbit Degree Vector), identical to ORCA (commonly though incorrectly called a GDV)\n"\
"	g = GDV (Graphlet Degree Vector) Note this is NOT what is commonly called a GDV, which is actually an ODV (above).\n"\
"	NOTE: the difference is that an ODV counts the number of nodes that touch all possible *orbits*, while a GDV lists\n"\
"		only the smaller vector of how many nodes touch each possible *graphlet* (independent of orbit).\n"\
"	i = {i}ndex: each line is a graphlet with columns: canonical ID, then k nodes in canonical order; useful since\n"\
"	    two lines with the same first column constitutes a PERFECT k-node local alignment between the two graphlets.\n"\
"	cX = Community detection where X is g for graphlet (the default) or o for orbit (which uses FAR more memory!)\n"\
"	    Each line consists of:   node (orbit|graphlet) count [TAB] neighbors.\n"\
"	r = index with {r}oot node orbit: each line is a canonical ID + the orbit of the root node, then k nodes in canonical order; produces better seeds when the index is queried by the alignment algorithm\n"\
"	d = graphlet neighbor {D}istribution\n"\
"    -d{displayMode: single character controls how canonical IDs are displayed. (DEFAULT=-mi): \n"\
"	o = ORCA numbering\n"\
"	j = JESSE numbering\n"\
"	b = explicit binary representation of the half-adjacency matrix of the canonical graphlet\n"\
"	d = decimal (base-10) integer representation of the above binary\n"\
"	i = integer ordinal = sorting the above integers and numbering them 0, 1, 2, 3, etc.\n"\
"Less Common OPTIONS:\n"\
"    -q quiet mode: suppress progress reports; -qq=supress all notes; -qqq=supress warnings (not recommended)\n"\
"    -w (EXPERIMENTAL) input network has edge weights in 3rd column; output currently not well-defined\n"\
"    -t N[:M]: use threading (parallelism); break the task up into N jobs (default 1) allowing at most M to run at one time.\n"\
"       M can be anything from 1 to a compile-time-specified maximum possible value (MAX_POSSIBLE_THREADS in blant.h),\n"\
"       but defaults to 4 to be conservative.\n"\
"    -r seed: pick your own random seed\n"\
"    -S is for SANITY TESTING ONLY! It turns off the de-biasing alpha multipliers and produces BIASED samples\n"\
"    -i FILENAME: file containing \"nodes of interest\"; every graphlet sampled will have at least one of these nodes\n"\
"    -W windowSize: DEPRECATED. (use '-h' option for more)\n"\
"	-p windowRepSamplingMethod: (DEPRECATED) one of the below\n"\
"	    MIN (Minimizer); MAX (Maximizer); DMIN (Minimizer With Distance); DMAX (Maximizer with Distance);\n"\
"	    LFMIN (Least Frequent Minimizer); LFMAX (Least Frequent Maximizer)\n"\
"	-P windowRepIterationMethods is one of: COMB (Combination) or DFS\n" \
"	-l windowRepLimitMethod is one of: [suffix N: limit to Top N satisfied graphlets]\n"\
"	    DEG (graphlet Total Degree); EDGE (1-step away numEdges)\n"\
"\nOPTIONS specific to -sINDEX sampling mode:\n" \
"   -M max multiplicity (INDEX sampling only: max allowed ambiguous permutations in indexed graphlets; M=0 means no max)\n" \
"   -T = top percent to expand to in -sINDEX sampling method (default 0)\n" \
"   -o = the orbit to use for the heuristic function\n" \
"   -f = the .orca4 file for the network\n" \
"   -a = sets whether ties are broken alphabetically or reverse alphabetically"\
;

static void SigEarlyAbort(int sig) {
    Warning("caught signal %d; setting _earlyAbort=true", sig);
    _earlyAbort = true;
}

// The main program, which handles multiple threads if requested.  We simply fire off a bunch of parallel
// blant *processes* (not threads, but full processes), and simply merge all their outputs together here
// in the parent.
int main(int argc, char *argv[])
{
    //printf("char %d short %d int %d long %d long long %d Gint %d\n", sizeof(char), sizeof(short), sizeof(int), sizeof(long), sizeof(long long), sizeof(Gint_type));
    // ENABLE_MEM_DEBUG(); // requires including "mem-debug.h" in blant.h (NOT at the top of blant.c!)
    int i, j, opt, multiplicity=1;
    unsigned long numSamples=0;
    double windowRep_edge_density = 0.0;
    int exitStatus = 0;

    assert(MAX_K <= TINY_SET_SIZE);

    if(argc == 1)
    {
	puts(USAGE_SHORT);
	exit(1);
    }

    signal(SIGUSR1, SigEarlyAbort);

    _numThreads = 1;
    _maxThreads = sysconf(_SC_NPROCESSORS_ONLN);

    _k = 0; _k_small = 0;

    int odv_fname_len = 0;

    // When adding new options, please insert them in ALPHABETICAL ORDER. Note that options that require arguments
    // (eg "-k3", where 3 is the argument) require a colon appended; binary options (currently only A, C and h)
    // have no colon appended.
    while((opt = getopt(argc, argv, "a:d:c:e:f:F:g:hi:k:K:l:M:m:n:o:P:p:qr:Rs:t:T:wW:x:X")) != -1)
    {
	switch(opt)
	{
	long nSampArg;
	case 'q': do ++_quiet; while(optarg && *optarg++);
	    break;
	case 'h':
	    printf("%s\n", USAGE_LONG);
	    #if __MINGW32__ || __WIN32__ || __CYGWIN__
	    printf("Note: current TSET size is %u bits\n", 8*sizeof(TSET));
	    #else
	    printf("Note: current TSET size is %lu bits\n", 8*sizeof(TSET));
	    #endif
	    exit(1); break;
	case 'i':
	    interestFile = fopen(optarg, "r");
	    if(!interestFile) Fatal("cannot open nodes-of-interest file '%s' while processing -i option", optarg);
	    // file will be read later, after we read the graph input file so we have the names
	    break;
	case 'F':
	    if(_freqDisplayMode != freq_display_mode_undef) Fatal("-C option cannot appear more than once");
	    switch (*optarg)
	    {
		case 'n': _freqDisplayMode = freq_display_mode_count; break;
		case 'c': _freqDisplayMode = freq_display_mode_concentration; break;
		case 'a': case 'e': _freqDisplayMode = freq_display_mode_estimate_absolute; break;
		default: Fatal("-C%c: unknown frequency display mode", *optarg); break;
	    }
	    break;
	case 'm':
	    if(_outputMode != undef) Fatal("tried to define output mode twice");
	    switch(*optarg)
	    {
	    case 'c': _outputMode |= communityDetection;
		switch(*(optarg+1)) {
		case 'o': _communityMode='o'; break;
		case '\0': case 'g': _communityMode='g'; break;
		default: Fatal("-mc%c: unknown community mode; valid values g or o\n", *(optarg+1)); break;
		}
		break;
	    case 'm': _outputMode |= indexMotifs; break;
	    case 'M': _outputMode |= indexMotifOrbits; break;
	    case 'i': _outputMode |= indexGraphlets; break;
	    case 'r': _outputMode |= indexGraphletsRNO; break;
	    case 'j': _outputMode |= indexOrbits; break;
	    case 'f': _outputMode |= graphletFrequency; break;
	    case 'g': _outputMode |= outputGDV; break;
	    case 'o': _outputMode |= outputODV; break;
	    case 'd': _outputMode |= graphletDistribution; break;
	    case 'p': _outputMode |= (predict|outputGDV|outputODV); break;
	    case 'q': _outputMode |= predict_merge; break;
	    default: Fatal("-m%c: unknown output mode \"%c\"", *optarg,*optarg);
	    break;
	    }
	    break;
	case 'd':
	    if (_displayMode != undefined) Fatal("tried to define canonical display mode twice");
	    switch(*optarg)
	    {
	    case 'b': _displayMode = binary; break;
	    case 'd': _displayMode = decimal; break;
	    case 'i': _displayMode = ordinal; break;
	    case 'j': _displayMode = jesse; break;
	    case 'o': _displayMode = orca; break;
	    case 'n': _displayMode = noncanonical; break;
	    default: Fatal("-d%c: unknown canonical display mode:n"
		    "\tmodes are i=integer ordinal, d=decimal, b=binary, o=orca, j=jesse", *optarg);
	    break;
	    }
	    break;
	case 't': 
        _numThreads = atoi(optarg);
        if(_numThreads > _maxThreads) Fatal("More threads specified than available on system.");
	    break;
	case 'r': _seed = atoi(optarg); if(_seed==-1)Apology("seed -1 ('-r -1' is reserved to mean 'uninitialized'");
	    break;
	case 'R': _rawCounts=true;
	    break;
	case 's':
	    if (_sampleMethod != -1) Fatal("Tried to define sampling method twice");
	    else if (strncmp(optarg, "NBE", 3) == 0)
		_sampleMethod = SAMPLE_NODE_EXPANSION;
	    else if (strncmp(optarg, "SEC", 3) == 0)
		_sampleMethod = SAMPLE_SEQUENTIAL_CHAINING;
	    else if (strncmp(optarg, "FAYE", 4) == 0) {
		if (strncmp(optarg, "FAYE!",5) != 0) Warning("FAYE is an ancient variant of NBE and produces counts with potentially large biases; suppress this warning by appending an exclamation mark");
		_sampleMethod = SAMPLE_FAYE;
	    }
	    else if (strncmp(optarg, "EBE", 3) == 0) {
		// if (strncmp(optarg, "EBE!",4) != 0) Warning("EBE is very fast on dense networks but produces counts with potentially extreme biases; suppress this warning by appending an exclamation mark");
		_sampleMethod = SAMPLE_EDGE_EXPANSION;
	    }
	    else if (strncmp(optarg, "MCMC",4) == 0) {
		//if (strncmp(optarg, "MCMC!",5) != 0) Warning("MCMC produces unbiased but high variance graphlet counts; suppress this warning by appending an exclamation mark");
		_sampleMethod = SAMPLE_MCMC;
		if (strchr(optarg, 'u') || strchr(optarg, 'U'))
		    _MCMC_EVERY_EDGE=true;
	    }
	    else if (strncmp(optarg, "EDGE_COVER", 10) == 0) {
		_sampleMethod = SAMPLE_MCMC;
		_sampleSubmethod = SAMPLE_MCMC_EC;
	    }
	    else if (strncmp(optarg, "RES", 3) == 0) {
		if (strncmp(optarg, "RES!",4) != 0) Warning("Reservoir sampling (RES) is unbiased but VERY slow; append an exclamation mark to supress this warning");
		_sampleMethod = SAMPLE_RESERVOIR;
	    }
	    else if (strncmp(optarg, "AR", 2) == 0){
		_sampleMethod = SAMPLE_ACCEPT_REJECT;
		if (strncmp(optarg, "AR!",3) != 0) Warning("Accept/Reject sampling (AR) is unbiased but EXTREMELY slow; append an exclamation mark to supress this warning");
	    }
	    else if (strncmp(optarg, "INDEX", 5) == 0)
		_sampleMethod = SAMPLE_INDEX;
	    else
	    {
		_sampleFileName = optarg;
		if(strcmp(optarg,"STDIN") == 0) _sampleFile = stdin;
		else _sampleFile = fopen(_sampleFileName, "r");
		if(!_sampleFile)
		    Fatal("Unrecognized sampling method '%s'; recognized options are NBE, MCMC, SEC, EBE, RES, FAYE, AR, or a filename (that can be 'STDIN'), but file '%s' cannot be opened", optarg, _sampleFileName);
		_sampleMethod = SAMPLE_FROM_FILE;
	    }
	    break;
	case 'P': _precisionMode = worst; // fall through, do not break
	case 'p':
	    if(atof(optarg) < 1) { // user has asked for relative precision
		_desiredPrec = atof(optarg);
		if(_desiredPrec <= 0)
		    Fatal("invalid requested precision %g must be in (0,1)", _desiredPrec);
		_desiredDigits = -log(_desiredPrec)/log(10);
	    }
	    else { // user has requested digits of precision
		_desiredDigits = atof(optarg);
		if(_desiredDigits <= 0) Fatal("invalid requested digits of precision %g must be > 0", _desiredDigits);
		_desiredPrec = pow(10, -_desiredDigits);
	    }
	    if(_desiredDigits > 3)
		Warning("requesting more than 3 digits of precision may be infeasible; you've requested %g", _desiredDigits);
	    if(_confidence) Fatal("Please specify confidence (-c option) AFTER specifying precision with -p or -P");
	    char wChar = optarg[strlen(optarg)-1];
	    if(isalpha(wChar)) {
		switch(wChar) {
		case 'w': case 'W': _precisionWt=PrecWtRaw; break;
		case 'l': case 'L': _precisionWt=PrecWtLog; break;
		default: Fatal("unknown precision weighting %c", wChar);
		}
	    }
        _stopMode = precision;
	    break;
	case 'c': _confidence = atof(optarg);
	    if(_confidence <= 0) Fatal("confidence must be in (0,1), not %g", _confidence);
	    if(_confidence >= 1) _confidence /= 100; // user specified percent
	    break;
	case 'k': _k = atoi(optarg);
	    if (_GRAPH_GEN && _k >= 33) {
		_k_small = _k % 10; // used in windowing code, obsolete
		if (!(3 <= _k_small && _k_small <= MAX_K))
		    Warning("%s\nk [%d] must be between 3 and %d\n%s", USAGE_SHORT, _k_small, MAX_K);
		_k /= 10;
		assert(_k_small <= _k);
	    } // First k indicates stamping size, second k indicates KS test size.
	    if (!(3 <= _k && _k <= MAX_K)) Warning("%s\nk [%d] must be between 3 and %d\n%s", USAGE_SHORT, _k, MAX_K);
	    break;
	case 'W': _window = true; _windowSize = atoi(optarg); break;
	case 'w': _weighted = true; break;
	case 'x':
	    if (_windowSampleMethod != -1) Fatal("Tried to define window sampling method twice");
	    else if (strncmp(optarg, "DMIN", 4) == 0)
		_windowSampleMethod = WINDOW_SAMPLE_MIN_D;
	    else if (strncmp(optarg, "DMAX", 4) == 0)
		_windowSampleMethod = WINDOW_SAMPLE_MAX_D;
	    else if (strncmp(optarg, "MIN", 3) == 0)
		_windowSampleMethod = WINDOW_SAMPLE_MIN;
	    else if (strncmp(optarg, "MAX", 3) == 0)
		_windowSampleMethod = WINDOW_SAMPLE_MAX;
	    else if (strncmp(optarg, "LFMIN", 5) == 0)
		_windowSampleMethod = WINDOW_SAMPLE_LEAST_FREQ_MIN;
	    else if (strncmp(optarg, "LFMAX", 5) == 0)
		_windowSampleMethod = WINDOW_SAMPLE_LEAST_FREQ_MAX;
	    else if (strncmp(optarg, "DEGMAX", 6) == 0)
		_windowSampleMethod = WINDOW_SAMPLE_DEG_MAX;
	    else
		Fatal("Unrecognized window searching method specified. Options are: -p[u|U]{MIN|MAX|DMIN|DMAX|LFMIN|LFMAX|DEGMAX}\n");
	    break;
	case 'X':
	    if (strncmp(optarg, "COMB", 4) == 0)
		_windowIterationMethod = WINDOW_ITER_COMB;
	    else if (strncmp(optarg, "DFS", 3) == 0)
		_windowIterationMethod = WINDOW_ITER_DFS;
	    else
		Fatal("Unrecognized window Iteration method specified. Options are: -P{COMB|DFS}\n");
	    break;
	case 'l':
	    if (_windowRep_limit_method != WINDOW_LIMIT_UNDEF) Fatal("Tried to define window limiting method twice");
	    if (strncmp(optarg, "n", 1) == 0 || strncmp(optarg, "N", 1) == 0) {
		_windowRep_limit_neglect_trivial = true; optarg += 1;
	    }
	    if (strncmp(optarg, "DEG", 3) == 0) {
		    _windowRep_limit_method = WINDOW_LIMIT_DEGREE; optarg += 3;
	    }
	    else if (strncmp(optarg, "EDGE", 4) == 0) {
		    _windowRep_limit_method = WINDOW_LIMIT_EDGES; optarg += 4;
	    }
	    else
		    Fatal("Unrecognized window limiting method specified. Options are: -l{DEG}{EDGE}{limit_num}\n");
	    _numWindowRepLimit = atoi(optarg);
	    if (!_numWindowRepLimit) {_numWindowRepLimit = 10; _numWindowRepArrSize = _numWindowRepLimit;}
	    _windowRep_limit_heap = HeapAlloc(_numWindowRepLimit, asccompFunc, NULL);
	    break;
	case 'n': nSampArg = atol(optarg);
	    if(nSampArg < 0) Fatal("%s\nFatal Error: numSamples [%s] must be a non-negative integer", USAGE_SHORT, optarg);
	    numSamples = nSampArg;
	    char lastChar = optarg[strlen(optarg)-1];
	    if(!isdigit(lastChar))
		switch(lastChar) {
		case 'b': case 'B': case 'g': case 'G': numSamples *= 1024; // do NOT break, fall through
		case 'm': case 'M': numSamples *= 1024; // do NOT break, fall through
		case 'k': case 'K': numSamples *= 1024; break;
		default: Fatal("%s\nERROR: numSamples can be appended by k, m, b, or g but not %c\n%s", USAGE_SHORT, lastChar);
        _stopMode = num_samples;
        _numSamples = numSamples;
		break;
	    }
	    //fprintf(stderr, "numSamples set to %d\n", numSamples);
	    break;
	case 'K': _KS_NUMSAMPLES = atoi(optarg);
	    break;
	case 'e':
	    _min_edge_count = _GRAPH_GEN_EDGES = atoi(optarg);
	    windowRep_edge_density = atof(optarg);
	    break;
	case 'g':
	    if (!GEN_SYN_GRAPH) Fatal("Turn on Global Variable GEN_SYN_GRAPH");
	    _GRAPH_GEN = true;
	    if (_genGraphMethod != -1) Fatal("Tried to define synthetic graph generating method twice");
	    else if (strncmp(optarg, "NBE", 3) == 0) _genGraphMethod = GEN_NODE_EXPANSION;
	    else if (strncmp(optarg, "MCMC", 4) == 0) Apology("MCMC for Graph Syn is not ready");  // _genGraphMethod = GEN_MCMC;
	    else Fatal("Unrecognized synthetic graph generating method specified. Options are: -g{NBE|MCMC}\n");
	    break;
	case 'M': multiplicity = atoi(optarg);
	    if(multiplicity < 0) Fatal("%s\nERROR: multiplicity [%d] must be non-negative\n", USAGE_SHORT, multiplicity);
	    break;
	case 'T': _topThousandth = atoi(optarg);
	    break;
	case 'o': _orbitNumber = atoi(optarg);
	    break;
	case 'f':
	    odv_fname_len = strlen(optarg);
	    _odvFile = malloc(sizeof(char) * odv_fname_len);
	    strncpy(_odvFile, optarg, odv_fname_len);
	    break;
	case 'a':
	    _alphabeticTieBreaking = (atoi(optarg) != 0);
	    break;
	default: Fatal("Run without command arguments for short usage message, or with -h for longer one");
	    break;
	}
    }


    if (_orbitNumber != -1) {
        if (_odvFile != NULL) {
            parseOdvFromFile(_odvFile);
        } else {
            Fatal("an ODV orbit number was provided, but no ODV file path was supplied");
        }
    }

    if (_sampleMethod == SAMPLE_INDEX && _k <= 5) Fatal("k is %d but must be at least 6 for INDEX sampling method because there are no unambiguous graphlets for k<=5",_k);

    if(_outputMode == undef) _outputMode = graphletFrequency; // default to frequency, which is simplest
    if(_freqDisplayMode == freq_display_mode_undef) _freqDisplayMode = freq_display_mode_estimate_absolute; // Default to estimating count

    if(numSamples && _desiredPrec) Fatal("cannot specify both -n (sample size) and -[Pp] (desired precision)");
    if(!numSamples && !_desiredPrec) { // default to 2 digits of precision at 99.9% confidence
	_desiredDigits = DEFAULT_DIGITS;
	_desiredPrec = pow(10, -_desiredDigits);
    }
    if (_sampleMethod == -1) {
	if(_desiredPrec)
	    _sampleMethod = SAMPLE_SEQUENTIAL_CHAINING; // MCMC samples are not independent, so use SEC for CI's
	else
	    _sampleMethod = SAMPLE_MCMC;
    }
    if(!_quiet && !(_outputMode & predict_merge)) Note("Sampling method is %s", SampleMethodStr());

    if(_desiredPrec && _confidence == 0)
	_confidence = (1-_desiredPrec/10);
    if(_desiredPrec && _sampleMethod == SAMPLE_MCMC)
	Warning("you've chosen MCMC sampling with confidence intervals; SEC is recommended since adjacent MCMC samples are not independent");

    FILE *fpGraph;
    int piped = 0;
    if(!argv[optind])
    {
	fpGraph = stdin;
	if(isatty(0) && !_quiet) Warning("reading graph input file from terminal, press ^D to finish");
    }
    else {
	char *graphFileName = argv[optind];
	fpGraph = readFile(graphFileName, &piped);
	if(!fpGraph) Fatal("cannot open graph input file '%s'\n", argv[optind]);
	optind++;
    }
    assert(optind == argc || _GRAPH_GEN || _windowSampleMethod == WINDOW_SAMPLE_DEG_MAX);

    SetBlantDirs(); // Needs to be done before reading any files in BLANT directory
    SetGlobalCanonMaps(); // needs _k to be set
    LoadMagicTable(); // needs _k to be set

    if (_window && _windowSize >= 3) {
        if (_windowSampleMethod == -1) Fatal("Haven't specified window searching method. Options are: -p{MIN|MAX|DMIN|DMAX|LFMIN|LFMAX}\n");
        if(_windowSize < _k) Fatal("windowSize must be at least size k\n");
        _MAXnumWindowRep = CombinChooseDouble(_windowSize, _k);
        _numWindowRepArrSize = _MAXnumWindowRep > 0 ? MIN(_numWindowRepArrSize, _MAXnumWindowRep) : _numWindowRepArrSize;
	// _windowReps needs true Calloc/Free since they may be realloc'd on-the-fly.
        _windowReps = Calloc(_numWindowRepArrSize, sizeof(int*));
        for(i=0; i<_numWindowRepArrSize; i++) _windowReps[i] = Calloc(_k+1, sizeof(int));
        if (windowRep_edge_density < 0) windowRep_edge_density = 0;
		if (windowRep_edge_density > 1) windowRep_edge_density = 1;
		_windowRep_min_num_edge = (int) CombinChooseDouble(_k, 2) * windowRep_edge_density;
		if (_windowRep_min_num_edge < 0) Fatal("WindowRep minimum number of edges must be larger than 0. Check edge density\n");
    }

    // This section of code computes the info necessary to implement "multiplicity" mode ('M' above), which controls
    // whether to output a sampled graphlet at all during INDEXING based on how much "ambiguity" there is in its
    // local alignment. The more permutations of the graphlet there are, the less useful it is for seeding local
    // alignments and therefore the less useful as a database index entry.
    if(_sampleMethod == SAMPLE_INDEX || _window) {
	_windowRep_allowed_ambig_set = SetAlloc(_numCanon);
	SET *orbit_temp = SetAlloc(_numOrbits);
	for(i=0; i<_numCanon; i++) {
	    if (SetIn(_connectedCanonicals, i)) {
		// calculate number of permutations for the given canonical graphlet
		// (loop through all unique orbits and count how many times they appear)
		// the formula is for every unique orbit, multiply the number of permutations by
		// the factorial of how many appearances that unique orbit has
		// if there is one orbit with three nodes and a second orbit with 2 nodes,
		// the number of permutations would be (3!)(2!)
		// NOTE: this is not the most efficient algorithm since it doesn't use hash tables.
		// I didn't want to overcomplicate it because it only happens once per run.
		// however, if speed is important (this currently takes about 5 seconds on k=8) this can be sped up
		for(j=0; j<_k; j++) SetAdd(orbit_temp, _orbitList[i][j]);
		unsigned uniq_orbits[_k];
		unsigned num_uniq_orbits = SetToArray(uniq_orbits, orbit_temp);
		unsigned uniq_orbit_i;
		unsigned total_orbit_perms = 1;

		for (uniq_orbit_i=0; uniq_orbit_i<num_uniq_orbits; uniq_orbit_i++) {
		    unsigned orbit_appearances = 0;
		    unsigned orbit_i = 0;

		    for (orbit_i=0; orbit_i<_k; orbit_i++) {
			if (_orbitList[i][orbit_i] == uniq_orbits[uniq_orbit_i]) {
			    orbit_appearances++;
			    total_orbit_perms *= orbit_appearances;
			}
		    }
		}

		// I know it's inefficient to put multiplicity here instead of around the whole orbit perm calculation code but
		// it increases readability, at least until orbit perm calculation is put into a function
		if(multiplicity == 0 || total_orbit_perms <= multiplicity) { // multiplicity = 0 means any ambiguity is allowed
		    SetAdd(_windowRep_allowed_ambig_set, i);
		}
		SetEmpty(orbit_temp);
	    }
	}
	SetFree(orbit_temp);
    } else {
	if(!numSamples && !_desiredPrec)
	    Fatal("must specify either desired precision using -[Pp] (preferred) or number of samples (less preferred)");
    }

    // Read network using native Graph routine.
    GRAPH *G = GraphReadEdgeList(fpGraph, SPARSE, _supportNodeNames, _weighted);
    if(_useComplement) G->useComplement = true;

    if(_supportNodeNames)
    {
	assert(G->name);
	_nodeNames = G->name;
    }
    if(fpGraph != stdin) closeFile(fpGraph, &piped);

    // Always allocate this set; if there are no "nodes of interest" then every node is a possible start done
    _startNodes = Calloc(G->n, sizeof(unsigned));
    _startNodeSet = SetAlloc(G->n);
    if(interestFile) {
	if(_sampleMethod != SAMPLE_NODE_EXPANSION) {
	    Warning("sampleMethod is being set to NBE to accomodate nodes-of-interest file");
	    _sampleMethod = SAMPLE_NODE_EXPANSION;
	}
	char nodeName[BUFSIZ];
	while(1 == fscanf(interestFile, "%s", nodeName)) {
	    foint nodeNum;
	    if(_supportNodeNames) {
                if(!BinTreeLookup(G->nameDict, (foint)nodeName, &nodeNum))
                    Fatal("nodes-of-interest file contains non-existent node '%s'", nodeName);
	    } else {
		nodeNum.i = atoi(nodeName);
		if(nodeNum.i < 0 || nodeNum.i >= G->n)
                    Fatal("nodes-of-interest file contains non-existent node '%d'", nodeNum.i);
	    }
	    if(SetIn(_startNodeSet, nodeNum.i))
		Fatal("nodes-of-interest cannot contain duplicate nodes but we've already seen '%s'", nodeName);
	    _startNodes[_numStartNodes] = nodeNum.i;
	    SetAdd(_startNodeSet, nodeNum.i);
	    ++_numStartNodes;
	    if(_numStartNodes > G->n)
		Fatal("nodes-of-interest file contains '%d' entries, which is more nodes (%d) than input graph",
		    _numStartNodes, G->n);
	}
    #if 0 // this isn't actually needed with NBE
	if(_numStartNodes < _k)
	    Fatal("nodes-of-interest file must contain at least k=%d entries, but contains only %d", _k, _numStartNodes);
    #endif
	Note("Read %d nodes-of-interest", _numStartNodes);
	assert(SetCardinality(_startNodeSet) == _numStartNodes);
    } else {
	int l;
	_numStartNodes = G->n;
	for(l=0; l<G->n; l++) {
	    _startNodes[l]=l;
	    SetAdd(_startNodeSet, l);
	}
    }

    if(_outputMode & communityDetection) {
	if(_communityMode == 'o' || _communityMode=='g') // allocate sets for [node][orbit], but 2nd dimension only when needed
	    _communityNeighbors = (SET***) Calloc(G->n, sizeof(SET**)); // elements are only allocated when needed
	else
	    Fatal("unknown _communityMode %c",_communityMode);
    }

    if (_windowSampleMethod == WINDOW_SAMPLE_DEG_MAX) {
        FILE *fp;
        _graphNodeImportance = Ocalloc(G->n, sizeof(float));
        if((optind + 1) == argc) {
            _supportNodeImportance = true;
            fp = fopen(argv[optind++], "r");
            if (fp == NULL) Fatal("cannot open graph Node Importance File.");
            char line[BUFSIZ], nodeName[BUFSIZ];
            foint nodeNum;
            float importance;

            while(fgets(line, sizeof(line), fp))
            {
                if(sscanf(line, "%s%f ", nodeName, &importance) != 2)
                    Fatal("GraphNodeImportance: Error while reading\n");
                if(!BinTreeLookup(G->nameDict, (foint)nodeName, &nodeNum))
                    Fatal("Node Importance Error: %s is not in the Graph file\n", nodeName);
                _graphNodeImportance[nodeNum.i] = importance;
            }
        }
    }

#if GEN_SYN_GRAPH
    FILE *fpSynGraph = NULL;
    if (_GRAPH_GEN) {
    	if (_k_small == 0) _k_small = _k;
        if (numSamples == 0) Fatal("Haven't specified sample size (-n sampled_size)");
        if (_confidence == 0) _confidence = 0.05;
        if (_KS_NUMSAMPLES == 0) _KS_NUMSAMPLES = 1000;
        if (_GRAPH_GEN_EDGES == 0) _GRAPH_GEN_EDGES = G->numEdges;
        if((optind + 1) == argc) {
            fpSynGraph = fopen(argv[optind++], "w");
            if (fpSynGraph == NULL) Fatal("cannot open synthetic graph outputfile.");
        }
        exitStatus = GenSynGraph(_k, _k_small, numSamples, G, fpSynGraph);
    }
#endif

    if(_outputMode & predict_merge)
	exitStatus = Predict_Merge(G);
    else
    exitStatus = RunBlantFromGraph(_k, numSamples, G);
    GraphFree(G);
    return exitStatus;
}
